### PR TITLE
[12.0] [ADD] account_payment_promissory_note: Module that adds a due date on payments

### PR DESCRIPTION
**Pull request by CarlosRoca13**
cc @Tecnativa it is necessary for this task TT18130

### CODE REVIEW

**Comment by CarlosRoca13** [comment]
Here you have an example of how it works
image

Please @sergio-teruel @carlosdauden, can you review it?

**Review by carlosdauden — APPROVED** [review]
LGTM :smiley:

**Review by sergio-teruel — APPROVED** [review]

**Comment by CarlosRoca13** [comment]
Fixed name @pedrobaeza

**Review by ernestotejeda — APPROVED** [review]

File: account_payment_promissory_note/README.rst [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+.. |badge4| image:: https://img.shields.io/badge/weblate-Translate%20me-F47D42.png
+    :target: https://translation.odoo-community.org/projects/account-payment-12-0/account-payment-12-0-account_payment_promissory_note
+    :alt: Translate me on Weblate
+.. |badge5| image:: https://img.shields.io/badge/runbot-Try%20me-875A7B.png
+    :target: https://runbot.odoo-community.org/runbot/96/12.0
+    :alt: Try me on Runbot
+
+|badge1| |badge2| |badge3| |badge4| |badge5| 
+
+This module allows you to set a date due in a payment.
```
Comment (pedrobaeza) [inline account_payment_promissory_note/README.rst]:
```suggestion
This module allows you to set a date due in a payment, as promissory notes require.
```

File: account_payment_promissory_note/README.rst [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+**Table of contents**
+
+.. contents::
+   :local:
+
+Usage
+=====
+
+#. Go to Invoicing > Customers / Vendors > Payments.
+#. Create a new one and select the deferred option.
```
Comment (pedrobaeza) [inline account_payment_promissory_note/README.rst]: Rename "Deferred" in UI by "Promissory note", and change documentation accordingly.

File: account_payment_promissory_note/README.rst [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+#. Now you can select a Date Due.
```
Comment (pedrobaeza) [inline account_payment_promissory_note/README.rst]:
```suggestion
#. Now you can select a date for the due of the promissory note.
```

File: account_payment_promissory_note/__manifest__.py [hunk 4]
```diff
@@ -0,0 +1,17 @@
+# Copyright 2018 Tecnativa - Carlos Dauden
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+
+{
+    'name': 'Account Payment Promissory Note',
+    'version': '12.0.1.0.0',
+    'license': 'AGPL-3',
+    'author': "Tecnativa,"
+              "Odoo Community Association (OCA)",
+    'category': 'Generic Modules/Accounting',
```
Comment (pedrobaeza) [inline account_payment_promissory_note/__manifest__.py]: Put `account` category of current version.

File: account_payment_promissory_note/__manifest__.py [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+    'website': "https://www.tecnativa.com",
```
Comment (pedrobaeza) [inline account_payment_promissory_note/__manifest__.py]:
```suggestion
    'website': "https://github.com/OCA/account-payment",
```

File: account_payment_promissory_note/models/account_payment.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+# Copyright 2018 Carlos Dauden <carlos.dauden@tecnativa.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import api, fields, models
+
+
+class AccountAbstractPayment(models.AbstractModel):
+    _inherit = 'account.abstract.payment'
+
+    deferred = fields.Boolean()
```
Comment (pedrobaeza) [inline account_payment_promissory_note/models/account_payment.py]:
```suggestion
    promissory_note = fields.Boolean()
```

File: account_payment_promissory_note/models/account_payment.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+            self.date_due = False
+            return
+        invoices = False
+        if self._name == 'account.payment':
+            invoices = self.invoice_ids
+        elif self._name == 'account.register.payments':
+            active_ids = self._context.get('active_ids')
+            invoices = self.env['account.invoice'].browse(active_ids)
+        if invoices:
+            self.date_due = max(invoices.mapped('date_due'))
```
Comment (pedrobaeza) [inline account_payment_promissory_note/models/account_payment.py]: Mention in the USAGE section about all the possible flows that I'm seeing here and also that the initial proposed due date is computed according this criteria.

File: account_payment_promissory_note/models/account_payment.py [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+    @api.onchange('deferred')
+    def _onchange_deferred(self):
+        if not self.deferred:
+            self.date_due = False
+            return
+        invoices = False
+        if self._name == 'account.payment':
+            invoices = self.invoice_ids
+        elif self._name == 'account.register.payments':
```
Comment (pedrobaeza) [inline account_payment_promissory_note/models/account_payment.py]: This mustn't be done here, but overwrite `_onchange_deferred` in the corresponding model, and perform this there.

File: account_payment_promissory_note/models/account_payment.py [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+            active_ids = self._context.get('active_ids')
+            invoices = self.env['account.invoice'].browse(active_ids)
+        if invoices:
+            self.date_due = max(invoices.mapped('date_due'))
+
+
+class AccountRegisterPayments(models.TransientModel):
```
Comment (pedrobaeza) [inline account_payment_promissory_note/models/account_payment.py]: Put it in a separate file and in its proper folder: `wizards`.

File: account_payment_promissory_note/models/account_payment.py [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+    _inherit = 'account.register.payments'
+
+    def get_payment_vals(self):
+        vals = super(AccountRegisterPayments, self).get_payment_vals()
```
Comment (pedrobaeza) [inline account_payment_promissory_note/models/account_payment.py]: Please add a test for covering this flow.

File: account_payment_promissory_note/models/account_payment.py [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+            return {'type': 'ir.actions.act_window_close'}
+        action = self.env.ref(
+            'account.action_account_payments_payable').read()[0]
+        action['views'] = [
+            (self.env.ref('account.view_account_payment_form').id, 'form')]
+        action['res_id'] = payment.id
+        return action
+
+
+class AccountPayment(models.Model):
```
Comment (pedrobaeza) [inline account_payment_promissory_note/models/account_payment.py:8]: Separate file

File: account_payment_promissory_note/views/account_payment_view.xml [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    <field name="inherit_id" ref="account.view_account_payment_invoice_form" />
+    <field name="arch" type="xml">
+        <xpath expr="//field[@name='payment_date']" position="after">
+            <field name="deferred"/>
+            <field name="date_due" attrs="{'invisible': [('deferred','=', False)]}"/>
+        </xpath>
+    </field>
+</record>
+
+<record id="view_account_payment_from_invoices" model="ir.ui.view">
```
Comment (pedrobaeza) [inline account_payment_promissory_note/views/account_payment_view.xml]: Separate file

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

**Comment by CarlosRoca13** [comment]
@pedrobaeza Changes done!

**Comment by CarlosRoca13** [comment]
@pedrobaeza, is this module is ready?

File: account_payment_promissory_note/models/account_abstract_payment.py [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+        string='Due Date',
+    )
+
+    @api.onchange('promissory_note')
+    def _onchange_promissory_note(self):
+        if not self.promissory_note:
+            self.date_due = False
+            return self.date_due
+        invoices = False
+        if self._name == 'account.payment':
```
Comment (pedrobaeza) [inline account_payment_promissory_note/models/account_abstract_payment.py]: This is still incorrect. You must override the method on `account.payment` model and add this logic

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

**Review by victoralmau — APPROVED** [review]

**Review by victoralmau — COMMENTED** [review]
Please, check Travis errors

**Comment by CarlosRoca13** [comment]
Solved @victoralmau

**Review by pedrobaeza — APPROVED** [review]

**Pull request merged**